### PR TITLE
Support for XNNPACK max pooling operator.

### PR DESCRIPTION
**Pull request by AshkanAliabadi**
Add support for XNNPACK 2D max pool operator.  The operator is enabled as a result of integration into at::max_pool2d(...), itself registered through native_functions.yaml.

Test Plan: CI

### CODE REVIEW

File: aten/src/ATen/native/xnnpack/Pooling.cpp [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+
+  TORCH_CHECK(
+      available(
+          input_channels,
+          output_channels,
+          kernel,
+          padding,
+          stride,
+          dilation,
+          ceil_mode,
```
Comment (kimishpatel) [inline aten/src/ATen/native/xnnpack/Pooling.cpp]: Is `ceil_mode` used somewhere by XNNPACK for maxpool op? PT documentation suggests that it is used to compute output shape, https://pytorch.org/docs/stable/nn.html#torch.nn.MaxPool2d. 
It it is not used then dont we need to figure out what output shape XNNPACK wants and be sure to provide buffer with that shape?
And also impose the restriction on whether maxpool2d can be mapped to xnnpack or not depending on the value of ceil_mode?

File: aten/src/ATen/native/Pooling.cpp [hunk 2]
```diff
@@ -132,6 +133,11 @@ Tensor max_pool2d(
     return at::mkldnn_max_pool2d(
         self, kernel_size, stride, padding, dilation, ceil_mode);
   }
+  if(xnnpack::use_max_pool2d(self, kernel_size, padding, stride,
```
Comment (kimishpatel) [inline aten/src/ATen/native/Pooling.cpp]: Why are we directly exposing this here? I thought, based on conv and linear, our philosophy was going to be not exposing this directly but require explicit opt-in via "xnnpackify.." short of transform of the network. @dreiss for comment.

Comment (kimishpatel) [reply aten/src/ATen/native/Pooling.cpp]: Also note that, this will also make it a little harder to fuse maxpool + relu.

**Review by kimishpatel — CHANGES_REQUESTED** [review]
Thanks Ashkan. Overall looks great. I have left a couple of comments.

File: aten/src/ATen/native/xnnpack/Pooling.cpp (hunk 1 above)
Comment (kimishpatel) [reply aten/src/ATen/native/xnnpack/Pooling.cpp]: I wonder if the seg faults in tests are related to the ceil_mode stuff.

File: aten/src/ATen/native/Pooling.cpp (hunk 2 above)
Comment (dzhulgakov) [reply aten/src/ATen/native/Pooling.cpp]: Binding it directly is good as it directly benefits existing models. It doesn't preclude us from doing more optimizations also like fusion you described.

**Review by dzhulgakov — COMMENTED** [review]
Looks good from overall point of view!

File: aten/src/ATen/native/xnnpack/Pooling.cpp [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+         // XNNPACK
+  return xnnpack::internal::available() &&
+         // Input
+         (input_channels > 0) &&
+         // Input == Output
+         (input_channels == output_channels) &&
+         // Kernel
+         ((1 <= kernel.size()) && (kernel.size() <= 2)) &&
+         (kernel[Layout::Parameter::height] > 0) &&
+         (kernel[Layout::Parameter::width] > 0) &&
```
Comment (ngimel) [inline aten/src/ATen/native/xnnpack/Pooling.cpp]: how does this work when kernel.size()==1? Layout::Parameter::width is 1, and you can't have kernel[1]. Ah, I see, you are expanding `kernel` in `create`, but then check for kernel.size() should be different.

File: aten/src/ATen/native/xnnpack/Pooling.cpp [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+namespace at {
+namespace native {
+namespace xnnpack {
+namespace internal {
+namespace max_pool2d {
+
+struct Context final {
+  Operator max_pool_op;
+
+  struct Output final {
```
Comment (ngimel) [inline aten/src/ATen/native/xnnpack/Pooling.cpp]: nit: you'll probably need similar structure for AvgPool and other kinds of pooling, so it makes sens not to put it in max_pool2d namespace

File: aten/src/ATen/native/Pooling.cpp (hunk 2 above)
Comment (dreiss) [reply aten/src/ATen/native/Pooling.cpp]: I agree.  With Dima.  We can have a separate diff that exposes the fused version.

File: aten/src/ATen/native/xnnpack/Convolution.cpp [hunk 5]
```diff
@@ -170,7 +170,7 @@ Tensor run(
       "XNNPACK Convolution not usable! "
       "Reason: The provided input tensor is either invalid or unsupported by XNNPACK.");
 
-  Tensor output = empty_with_tail_padding(
+  Tensor output_nhwc = empty_with_tail_padding(
```
Comment (dreiss) [inline aten/src/ATen/native/xnnpack/Convolution.cpp]: These changes should probably be a separate diff.

File: aten/src/ATen/native/xnnpack/Pooling.cpp [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+    const IntArrayRef padding,
+    const IntArrayRef stride,
+    const IntArrayRef dilation,
+    const bool ceil_mode,
+    const float output_min,
+    const float output_max) {
+  using namespace internal;
+
+  const int64_t input_channels = input.size(Layout::Activation4D::channels);
+  const int64_t output_channels = input_channels;
```
Comment (dreiss) [inline aten/src/ATen/native/xnnpack/Pooling.cpp]: If output channels is always equal to input channels, and you assert later that it always will be equal to input channels, why have it as a separate parameter?

File: aten/src/ATen/native/xnnpack/Pooling.cpp [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+}
+
+Tensor run(
+    const Context& context,
+    const Tensor& input) {
+  using namespace internal;
+
+  TORCH_CHECK(
+      usable(input, context.output.height, context.output.width),
+      "mobile::cpu::max_pool not usable!");
```
Comment (dreiss) [inline aten/src/ATen/native/xnnpack/Pooling.cpp]: Message needs to be updated.  Also, shouldn't this be checked when the context is created?

File: aten/src/ATen/native/xnnpack/Pooling.cpp [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+
+  const Tensor input_nhwc = input.contiguous(MemoryFormat::ChannelsLast);
```
Comment (dreiss) [inline aten/src/ATen/native/xnnpack/Pooling.cpp]: Do we need to check that input_nhwc was created with a guarding allocator, or is that only needed for the output?

File: aten/src/ATen/native/xnnpack/Pooling.cpp [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+
+  const xnn_status run_status = xnn_run_operator(
+      context.max_pool_op.get(),  // operator
+      nullptr);                   // threadpool
+
+  TORCH_INTERNAL_ASSERT(
+      xnn_status_success == run_status,
+      "xnn_run_operator failed!");
+
+  return output_nhwc.contiguous(input.suggest_memory_format());
```
Comment (dreiss) [inline aten/src/ATen/native/xnnpack/Pooling.cpp]: Why not just return output_nhwc?

File: aten/src/ATen/native/xnnpack/Pooling.cpp [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+    const IntArrayRef kernel,
+    const IntArrayRef padding,
+    const IntArrayRef stride,
+    const IntArrayRef dilation,
+    const bool ceil_mode) {
+  using namespace internal;
+
+  const int64_t input_channels = (input.dim() == 4) ?
+                                 input.size(Layout::Activation4D::channels) :
+                                 -1 /* Setting to a negative number to fail avalable() */;
```
Comment (dreiss) [inline aten/src/ATen/native/xnnpack/Pooling.cpp]: This seems like too tight of a coupling.  Can you add a separate check for dim==4 instead?

**Review by dreiss — CHANGES_REQUESTED** [review]
Integration looks good.  Should have some unit tests added for this.

File: aten/src/ATen/native/xnnpack/Convolution.cpp (hunk 5 above)
Comment (kimishpatel) [reply aten/src/ATen/native/xnnpack/Convolution.cpp]: Yes please. Can you remove these changes? Else I or you will run into some merge conflict.

File: aten/src/ATen/native/xnnpack/Pooling.cpp (hunk 3 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/xnnpack/Pooling.cpp]: I'm sorry I didn't get your point about kernel.size() being different, can you elaborate?  Yes, I'm allowing 1 and 2 but expanding to 2 prior to use.

File: aten/src/ATen/native/xnnpack/Pooling.cpp (hunk 9 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/xnnpack/Pooling.cpp]: I think PyTorch's convention is to return tensors in the same layout they came in so we'll have to switch back to NCHW if that's the layout the input tensor was in.  Dima / Natalia can confirm.

**Comment by AshkanAliabadi** [comment]
Thanks for the comments.  Will address and upload a new patch.

File: aten/src/ATen/native/xnnpack/Pooling.cpp (hunk 1 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/xnnpack/Pooling.cpp]: Right, maybe.  I'll investigate, thanks.

File: aten/src/ATen/native/xnnpack/Convolution.cpp (hunk 5 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/xnnpack/Convolution.cpp]: OK, no worries, will revert.

File: aten/src/ATen/native/xnnpack/Pooling.cpp (hunk 3 above)
Comment (ngimel) [reply aten/src/ATen/native/xnnpack/Pooling.cpp]: If you require 2, then you should not allow 1, and looks like you require 2, because otherwise kernel[1] will segfault.

Comment (AshkanAliabadi) [reply aten/src/ATen/native/xnnpack/Pooling.cpp]: Aaah ... how could I have missed that?!  :/  Thanks.

File: aten/src/ATen/native/xnnpack/Pooling.cpp (hunk 9 above)
Comment (dreiss) [reply aten/src/ATen/native/xnnpack/Pooling.cpp]: If that's the convention, then it makes sense to stick with it.

We should make sure that if we're returning NHWC, we're doing so with a guarding allocator to get maximum performance from a sequence of XNNPACK ops.

File: aten/src/ATen/native/xnnpack/Pooling.cpp (hunk 6 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/xnnpack/Pooling.cpp]: Removed.

File: aten/src/ATen/native/xnnpack/Pooling.cpp (hunk 7 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/xnnpack/Pooling.cpp]: Updated the message.

The input tensor is not available at creation time.  available() checks any parameter provided at creation time while usable gates anything that depends on the input tensor.

File: aten/src/ATen/native/xnnpack/Pooling.cpp (hunk 8 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/xnnpack/Pooling.cpp]: Yes, this should use allocate_padded_if_needed() in Kimish's patch: only re-allocate input if not already allocated with this allocator.

File: aten/src/ATen/native/xnnpack/Pooling.cpp (hunk 10 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/xnnpack/Pooling.cpp]: Done.

File: aten/src/ATen/native/xnnpack/Pooling.cpp (hunk 4 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/xnnpack/Pooling.cpp]: Thanks, I'll move it in a future patch after Kimish merges his.  The only place I can currently put it right now is inside Common.h which is starting to become a cluttered mess.

File: aten/src/ATen/native/Pooling.cpp (hunk 2 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/Pooling.cpp]: Great.

File: aten/src/ATen/native/xnnpack/Pooling.cpp (hunk 9 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/xnnpack/Pooling.cpp]: contiguous is a no-op if memory is already in the requested layout.  In other words, if the input tensor is in NHWC, contiguous will short circuit..

File: aten/src/ATen/native/xnnpack/Factory.cpp [hunk 11]
```diff
@@ -27,7 +29,9 @@ Tensor empty_with_tail_padding(
           },
           DispatchKeySet{DispatchKey::CPUTensorId}));
 
-  return tensor.resize_(size, memory_format);
+  return namedinference::propagate_names_if_nonempty(
+      tensor.resize_(size, memory_format),
+      maybe_names);
```
Comment (AshkanAliabadi) [inline aten/src/ATen/native/xnnpack/Factory.cpp]: Note: Had to add this to prevent a test in test/test_namedtensor.py from failing.  Said test expects tensor names to propagate.

File: aten/src/ATen/native/xnnpack/Pooling.cpp (hunk 4 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/xnnpack/Pooling.cpp]: OK, done now that Kimish's patch is merged.

File: aten/src/ATen/native/xnnpack/MaxPooling.cpp [hunk 12]
```diff
@@ -0,0 +1,367 @@
...
+         true;
+}
+
+Tensor run(
+    const Context& context,
+    const Tensor& input) {
+  using namespace internal;
+
+  const Tensor input_nhwc = input.contiguous(MemoryFormat::ChannelsLast);
+  const Tensor padded_input_nhwc = allocate_padded_if_needed(input_nhwc);
```
Comment (dreiss) [inline aten/src/ATen/native/xnnpack/MaxPooling.cpp]: This is potentially two copies.  Can we combine these into one?

File: aten/src/ATen/native/xnnpack/MaxPooling.cpp [hunk 13]
```diff
@@ -0,0 +1,367 @@
...
+  return max_pool2d::available(
+        input.size(Layout::Activation4D::channels),
+        parameters.kernel,
+        parameters.padding,
+        parameters.stride,
+        parameters.dilation,
+        ceil_mode,
+        internal::max_pool2d::Context::kMin,
+        internal::max_pool2d::Context::kMax) &&
+     max_pool2d::usable(
```
Comment (dreiss) [inline aten/src/ATen/native/xnnpack/MaxPooling.cpp]: Why do we need available and usable as separate functions?

File: aten/src/ATen/native/xnnpack/MaxPooling.cpp [hunk 14]
```diff
@@ -0,0 +1,367 @@
...
+          input.size(Layout::Activation4D::channels),
+          kernel,
+          padding,
+          stride,
+          dilation,
+          ceil_mode,
+          output_min,
+          output_max),
+      input);
+}
```
Comment (dreiss) [inline aten/src/ATen/native/xnnpack/MaxPooling.cpp]: This seems like an unnecessary layer of abstraction since we're not persisting the Context.  If you inline both create and run directly into `max_pool2d`, can you eliminate Context and shorten/simplify the entire diff?

**Review by dreiss — CHANGES_REQUESTED** [review]
Can you separate the named tensor and formatting/comment changes into separate diffs?

File: aten/src/ATen/native/xnnpack/MaxPooling.cpp (hunk 14 above)
Comment (AshkanAliabadi) [reply aten/src/ATen/native/xnnpack/MaxPooling.cpp]: Don't we want to maintain the ability to separate create and run in the future?

Comment (dreiss) [reply aten/src/ATen/native/xnnpack/MaxPooling.cpp]: Why?  We don't expect it will ever have a significant perf improvement over this implementation, right?

**Comment by AshkanAliabadi** [comment]
Breaking PR into smaller chunks per David's request. #35354. Closing.
